src/render/draw_heatmap.js

```diff
@@ -0,0 +1,137 @@
+// @flow
+
+const browser = require('../util/browser');
+const EXTENT = require('../data/extent');
+const mat4 = require('@mapbox/gl-matrix').mat4;
+const Texture = require('./texture');
+
+import type Painter from './painter';
+import type SourceCache from '../source/source_cache';
+import type HeatmapStyleLayer from '../style/style_layer/heatmap_style_layer';
+import type HeatmapBucket from '../data/bucket/heatmap_bucket';
+import type TileCoord from '../source/tile_coord';
+
+module.exports = drawHeatmap;
+
+function drawHeatmap(painter: Painter, sourceCache: SourceCache, layer: HeatmapStyleLayer, coords: Array<TileCoord>) {
+    if (painter.isOpaquePass) return;
+
+    const gl = painter.gl;
+
+    painter.setDepthSublayer(0);
+    painter.depthMask(false);
+
+    // Allow circles to be drawn across boundaries, so that
+    // large circles are not clipped to tiles
+    gl.disable(gl.STENCIL_TEST);
+
+    renderToTexture(gl, painter, layer);
```

Comment:
Can we please move this call to a different render pass to improve performance? Extrusion fills use a `3d` render pass to render the texture before rendering to the main FBO. See https://github.com/mapbox/mapbox-gl-native/issues/9286 for a description of this optimization, and https://github.com/mapbox/mapbox-gl-js/pull/5101 for @lbud implemented this for extrusion fills.

Reply:
@kkaefer @lbud should we call this pass differently then? e.g. `prerender` or something?

Reply:
Yes, @mollymerp and I were just talking about this yesterday, as hillshades use a pre-render pass as well — it would make sense to rename the `3d` pass to `prerender` and carefully move relevant parts of the current `3d` pass (I suppose for now we could move them into `draw_fill_extrusion.js`, although eventually if we add other 3D types it'll be moved to a file where it can be shared between types).